Add native commutators to PauliWord and PauliSentence classes

Qottmann opened the pull request:
Adding native commutators for paulis.
Works between `PauliSentence` and any of `Operator, PauliWord, PauliSentence`

Builds on top of https://github.com/PennyLaneAI/pennylane/pull/5051

Todo

- [x] Basic functionality
- [x] Basic functionality tests
- [x] update qml.commutator
- [x] changelog

Qottmann commented:
[sc-53927]

Qottmann commented:
Why native commutator support you ask?

```
n = 10
H1 = PauliWord({0:"X", 1:"Z"})
H2 = PauliWord({0:"X", 1:"Z"})
for i in range(n):
    H1 += 0.5 * PauliWord({i+1:"Y", i+2:"Z"})
    H2 += 0.5 * PauliWord({i+1:"X", i+2:"Y"})
>>> %timeit H1 | H2
253 µs ± 4.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit qml.commutator(H1, H2)
3.12 ms ± 348 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

mudit2812 commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -592,6 +631,34 @@ def __truediv__(self, other):
+        for pw1 in self:
+            for pw2 in other:
+                comm_pw, coeff = pw1._commutator(pw2)
+                if comm_pw is not None:
```
`_commutator` never returns `None`

```suggestion
                if len(comm_pw) != 0:
```

Qottmann replied:
Oh good catch! A relic from an old version :)

mudit2812 commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -592,6 +631,34 @@ def __truediv__(self, other):
+        return self.commutator(other)
+
+    def __ror__(self, other):
+        return -1 * qml.commutator(self, other, pauli=True)
```
Why not this?
```suggestion
        return -1 * self.commutator(other)
```

mudit2812 reviewed (commented):
Looks good to me! Looks great to me! You can also update `qml.commutator` now to use `PauliSentence.commutator` for `pauli=True`. I'm happy with the current changes and am ready to approve once `qml.commutator` is updated.

Qottmann commented:
Now fully removed the infix commutator, which removes the capability to have the first input be an Operator (e.g. `PauliX(0) | PauliWord({0:"Y"})`, which required some changes to the tests.

mudit2812 commented:
> Why native commutator support you ask?
> 
> ```
> n = 10
> H1 = PauliWord({0:"X", 1:"Z"})
> H2 = PauliWord({0:"X", 1:"Z"})
> for i in range(n):
>     H1 += 0.5 * PauliWord({i+1:"Y", i+2:"Z"})
>     H2 += 0.5 * PauliWord({i+1:"X", i+2:"Y"})
> >>> %timeit H1 | H2
> 253 µs ± 4.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
> >>> %timeit qml.commutator(H1, H2)
> 3.12 ms ± 348 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
> ```
> 
> 🚀

Considering the speedup, `qml.comm` should probably be updated to use the native commutators if `Pauli=True`.

Qottmann commented:
> Considering the speedup, qml.comm should probably be updated to use the native commutators if Pauli=True.

Is there something missing in https://github.com/PennyLaneAI/pennylane/pull/5052/commits/ba5166fd74d3387f5fba4bdbaced7c6776ef68cd ?

mudit2812 commented:
> > Considering the speedup, qml.comm should probably be updated to use the native commutators if Pauli=True.
> 
> Is there something missing in [ba5166f](https://github.com/PennyLaneAI/pennylane/commit/ba5166fd74d3387f5fba4bdbaced7c6776ef68cd) ?

Missed that change. Sorry about that!

mudit2812 commented on doc/releases/changelog-dev.md:
````diff
@@ -39,6 +40,16 @@
   + 2j * X(0) @ Z(1)
   ```
 
+  We can also compute commutators with Pauli operators natively with the `|` (or) operator.
+  ```pycon
+  >>> op1 = PauliWord({0:"X", 1:"X"})
+  >>> op2 = PauliWord({0:"Y"}) + PauliWord({1:"Y"})
+  >>> op1 | op2
+  2j * Z(0) @ X(1)
+  + 2j * X(0) @ Z(1)
+  ```
````
This example needs to be updated.

Qottmann replied:
Nice catch, thanks! ✅

mudit2812 reviewed (commented):
Thanks Korbinian. I'm ready to approve once test coverage is fixed 😄

Qottmann commented:
Realized there was a legacy clause in `PauliSentence.commutator` returning `copy(self)` when `len(other)=0`, which doesnt make sense as the empty PauliSentence is the 0 element.

Removed that bit, updated a test checking that the logic is sound now, and updated the docs.

Qottmann commented:
Ready for review now after some CI trouble.

This PR adds a commutator method native to PauliWord/Sentence instances that is substantially faster.
It also updates`qml.commutator` to use the native method `pauli1.commutator(pauli2)`.

albi3ro commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -312,6 +320,44 @@ def __truediv__(self, other):
+            return PauliSentence({new_word: coeff})
+
+        if isinstance(other, qml.operation.Operator):
+            op_self = qml.pauli.pauli_sentence(self)
```
Why not just `PauliSentence({self, 1})`? Or just reproducing some of that logic here?

Qottmann replied:
Yes good point, updated that ✅

albi3ro commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -484,6 +530,24 @@ class PauliSentence(dict):
+    Or, alternatively, use :func:`~commutator`.
+    >>> qml.commutator(op1, op2, pauli=True)
+
+    Note that we need to specify `pauli=True` as the function returns PennyLane operators by default.
```
```suggestion
    Note that we need to specify ``pauli=True`` as :func:`~.commutator` returns PennyLane operators by default.
```

albi3ro commented on pennylane/pauli/pauli_arithmetic.py:
```diff
@@ -312,6 +320,44 @@ def __truediv__(self, other):
+        return new_word, 2 * coeff
+
+    def commutator(self, other):
+        """comm between two PauliWords"""
```
I think the docstrings for `commutator` are worth fleshing out into our standard full docstring format.

Qottmann replied:
Added ✅ also for PauliSentence.commutator()

albi3ro reviewed (approved):
🎉

vincentmr reviewed (approved):
Nice work @Qottmann . Looks good to me.
